docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,275 @@
+
+```c++
+typedef struct _blob {
+    void* data;
+    size_t size;
+} blob_t;
+
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check0(void* ptr) {
+
+    blob_t* blob;
+
+    if (ptr == nullptr)
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // blobs to be rendered are limited to 4K
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+
+## 1.1 Ensuring memory is where it should be
+
+Remember that the code we are focused on is running _inside_ of the enclave. The host caller invoked this function from _outside_ the enclave, passing in a pointer to host memory containing well-formed data (in the non-malicious case). But in the Open Enclave security model, the host caller is untrusted - we must treat untrusted input with caution. 
+
+One important security check is to ensure that memory blocks are on the correct side of the security boundary. It's common for host applications to pass host memory blocks to the enclave, as is the case with this example code. The Open Enclave SDK provides two library functions for checking memory buffers: `oe_is_within_enclave()` and `oe_is_outside_enclave()`.
+
+Let's update the code to ensure that the memory block is at least the size of a `blob_t` and is located _strictly_ outside the secured enclave memory region before operating on it by calling `oe_is_outside_enclave()`.
+
+```diff
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check1(void* ptr) {
+
+    blob_t* blob;
+
+-   if (ptr == nullptr)
+-       return -1;
++   // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
++   if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
++       return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Blobs to be rendered are limited to 4K.
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+## 1.2 SGX enclaves: oe_is_within_enclave vs. oe_is_outside_enclave
+
+Let's highlight a logical mistake that we've seen made with SGX-based enclaves. Some have regarded the memory validation functions as boolean opposites of each other, that is: `oe_is_within_enclave() == !oe_is_outside_enclave()`. This is incorrect for several reasons.
+
+To understand why this function pair are _not_ boolean opposites, it helps to consider the SGX implementation: [memory.c](https://github.com/openenclave/openenclave/blob/08ebe60c1d2e3ea24ac634c673a420296bff3352/enclave/core/sgx/memory.c). There are three conditions that the functions validate:
+1)	The pointer is not null.
+2)	The bounding arithmetic operations do not wrap (i.e. numerical overflow).
```

Comment:
```suggestion
2)	The bounding arithmetic operations do not wrap (i.e., numerical overflow).
```
nit: always put comma after "i.e." per Chicago Manual of Style

Reply:
Yes! The Chicago Manual of Style.
The Microsoft Style Guide also states not to use "i.e.": https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/ie